In [1]:
#Magic command para importar arquivos ipynb
%run dados.ipynb


    
---
# Contexto
---
## Enunciado do Problema

Analise do Comportamento do Consumidor é uma análise detalhada do consumidor ideal de uma empresa. Ajuda o negócio a compreender seus consumidores e faz com que seja fácil para eles modificar os produtos de acordo com as necessidade especificas, comportamentos e preocupações de diferentes tipos de consumidores.

Analise Comportamento do consumidor ajuda um negócio a modificar seus produtos baseado no público-alvo de diferentes segmentos de consumidores. Por exemplo, no lugar de investir dinheiro para divulgar um novo produto para todos os consumidores que estão no bancos de dados da empresa, pode-se fazer uma análise de qual segmento de consumidores provavelmente compraria o produto e então divulgar o produto somente para esse segmento específico.


## Conteúdo

### Colunas

#### Pessoas
* ID: Identificador único do consumidor
* AnoNascimento: Ano de nascimento do consumidor
* GrauInstrução: Grau de instrução do consumidor
* EstadoCivil: Estado civil do consumidor
* Renda: Renda familiar anual do consumidor
* NumCrianca: Número de crianças que moram com o consumidor
* NumAdolescente: Número de adolescentes que moram com o consumidor
* DataConta: Data que cliente se registrou na companhia
* UltimaCompra: Número de dias desde a última compra
* Reclamacao: 1 se o consumidor reclamou nos últimos 2 anos, 0 se não


#### Produtos

* QtdVinhos: Quantidade gasta em vinho nos últimos 2 anos
* QtdFrutas: Quantidade gasta em frutas nos últimos 2 anos
* QtdCarne: Quantidade gasta em carnes nos últimos 2 anos
* QtdPeixe: Quantidade gasta em peixe nos últimos 2 anos
* QtdDoces: Quantidade gasta em doces nos últimos 2 anos
* QtdOuro: Quantidade gasta em ouro nos últimos 2 anos


#### Promoções

* QtdComprasDesconto:Número de pedidos feitos com desconto
* AceitouCam1: 1 se cliente aceitou a oferta na primeira campanha, 0 se não
* AceitouCam3:  1 se cliente aceitou a oferta na segunda campanha, 0 se não
* AceitouCam3:  1 se cliente aceitou a oferta na terceira campanha, 0 se não
* AceitouCam3:  1 se cliente aceitou a oferta na quarta campanha, 0 se não
* AceitouCam3:  1 se cliente aceitou a oferta na quinta campanha, 0 se não
* UltimaCampanha: 1 se cliente aceitou a oferta na última campanha, 0 se não

#### Onde comprou

* QtdComprasWeb: Número de compras feitas pela internet
* QtdCompraCatalogo: Número de compras efetuadas por catalogo
* QtdCompraLoja: Número de compras feitas na loja
* NumVisitasSite: Número de visitas feitas ao site da companhia no último mês


---
#### Dataframe definido no arquivo dados.ipynb e será utilizado para limpeza e tradução dos dados
---

In [2]:
df.head()

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                        
5524        1957  Graduation         Single  58138.0        0         0   
2174        1954  Graduation         Single  46344.0        1         1   
4141        1965  Graduation       Together  71613.0        0         0   
6182        1984  Graduation       Together  26646.0        1         0   
5324        1981         PhD        Married  58293.0        1         0   

     Dt_Customer  Recency  MntWines  MntFruits  ...  NumWebVisitsMonth  \
ID                                              ...                      
5524  04-09-2012       58       635         88  ...                  7   
2174  08-03-2014       38        11          1  ...                  5   
4141  21-08-2013       26       426         49  ...                  4   
6182  10-02-2014       26        11          4  ...                  6   
5324  19-01-2014       94       173         43  ...                  5   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
ID                                                                           
5524             0             0             0             0             0   
2174             0             0             0             0             0   
4141             0             0             0             0             0   
6182             0             0             0             0             0   
5324             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
ID                                                  
5524         0              3         11         1  
2174         0              3         11         0  
4141         0              3         11         0  
6182         0              3         11         0  
5324         0              3         11         0  

[5 rows x 28 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2240 entries, 5524 to 9405
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Year_Birth           2240 non-null   int64  
 1   Education            2240 non-null   object 
 2   Marital_Status       2240 non-null   object 
 3   Income               2216 non-null   float64
 4   Kidhome              2240 non-null   int64  
 5   Teenhome             2240 non-null   int64  
 6   Dt_Customer          2240 non-null   object 
 7   Recency              2240 non-null   int64  
 8   MntWines             2240 non-null   int64  
 9   MntFruits            2240 non-null   int64  
 10  MntMeatProducts      2240 non-null   int64  
 11  MntFishProducts      2240 non-null   int64  
 12  MntSweetProducts     2240 non-null   int64  
 13  MntGoldProds         2240 non-null   int64  
 14  NumDealsPurchases    2240 non-null   int64  
 15  NumWebPurchases      2240 non-null 

In [4]:
df.columns

Index(['Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

In [5]:
df.shape

(2240, 28)

In [6]:
df.describe()

Year_Birth         Income      Kidhome     Teenhome      Recency  \
count  2240.000000    2216.000000  2240.000000  2240.000000  2240.000000   
mean   1968.805804   52247.251354     0.444196     0.506250    49.109375   
std      11.984069   25173.076661     0.538398     0.544538    28.962453   
min    1893.000000    1730.000000     0.000000     0.000000     0.000000   
25%    1959.000000   35303.000000     0.000000     0.000000    24.000000   
50%    1970.000000   51381.500000     0.000000     0.000000    49.000000   
75%    1977.000000   68522.000000     1.000000     1.000000    74.000000   
max    1996.000000  666666.000000     2.000000     2.000000    99.000000   

          MntWines    MntFruits  MntMeatProducts  MntFishProducts  \
count  2240.000000  2240.000000      2240.000000      2240.000000   
mean    303.935714    26.302232       166.950000        37.525446   
std     336.597393    39.773434       225.715373        54.628979   
min       0.000000     0.000000         0.000000         0.000000   
25%      23.750000     1.000000        16.000000         3.000000   
50%     173.500000     8.000000        67.000000        12.000000   
75%     504.250000    33.000000       232.000000        50.000000   
max    1493.000000   199.000000      1725.000000       259.000000   

       MntSweetProducts  ...  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
count       2240.000000  ...        2240.000000   2240.000000   2240.000000   
mean          27.062946  ...           5.316518      0.072768      0.074554   
std           41.280498  ...           2.426645      0.259813      0.262728   
min            0.000000  ...           0.000000      0.000000      0.000000   
25%            1.000000  ...           3.000000      0.000000      0.000000   
50%            8.000000  ...           6.000000      0.000000      0.000000   
75%           33.000000  ...           7.000000      0.000000      0.000000   
max          263.000000  ...          20.000000      1.000000      1.000000   

       AcceptedCmp5  AcceptedCmp1  AcceptedCmp2     Complain  Z_CostContact  \
count   2240.000000   2240.000000   2240.000000  2240.000000         2240.0   
mean       0.072768      0.064286      0.013393     0.009375            3.0   
std        0.259813      0.245316      0.114976     0.096391            0.0   
min        0.000000      0.000000      0.000000     0.000000            3.0   
25%        0.000000      0.000000      0.000000     0.000000            3.0   
50%        0.000000      0.000000      0.000000     0.000000            3.0   
75%        0.000000      0.000000      0.000000     0.000000            3.0   
max        1.000000      1.000000      1.000000     1.000000            3.0   

       Z_Revenue     Response  
count     2240.0  2240.000000  
mean        11.0     0.149107  
std          0.0     0.356274  
min         11.0     0.000000  
25%         11.0     0.000000  
50%         11.0     0.000000  
75%         11.0     0.000000  
max         11.0     1.000000  

[8 rows x 25 columns]

---
#### Tradução das colunas
---
* Year_Birth: **AnoNascimento**
* Education: **GrauInstrução**
* Marital_Status: **EstadoCivil**
* Income: **Renda**
* Kidhome: **NumCrianca**
* Teenhome: **NumAdolescente**
* Dt_Customer: **DataConta**
* Recency: **UltimaCompra**
* MntWines: **QtdVinho**
* MntFruits: **QtdFrutas**
* MntMeatProducts: **QtdCarne**
* MntFishProducts: **QtdPeixe**
* MntSweetProducts: **QtdDoces**
* MntGoldProds: **QtdOuro**
* NumDealsPurchases: **QtdComprasDesconto**
* NumWebPurchases: **QtdComprasWeb**
* NumCatalogPurchases: **QtdCompraCatalogo**
* NumStorePurchases: **QtdCompraLoja**
* NumWebVisitsMonth: **NumVisitasSite**
* AcceptedCmp3: **AceitouCam3**
* AcceptedCmp4: **AceitouCam4**
* AcceptedCmp5: **AceitouCam5**
* AcceptedCmp1: **AceitouCam1**
* AcceptedCmp2: **AceitouCam2**
* Complain: **Reclamacao**
* Z_CostContact: **CustoContato**
* Z_Revenue: **Receita**
* Response: **UltimaCampanha**

In [7]:
novo=['AnoNascimento', 'GrauInstrução', 'EstadoCivil', 'Renda', 'NumCrianca', 'NumAdolescente', 'DataConta', 'UltimaCompra', 'QtdVinho', 'QtdFrutas',
     'QtdCarne', 'QtdPeixe', 'QtdDoces', 'QtdOuro', 'QtdComprasDesconto','QtdComprasWeb', 'QtdCompraCatalogo', 'QtdCompraLoja','NumVisitasSite','AceitouCam3',
      'AceitouCam4', 'AceitouCam5', 'AceitouCam1', 'AceitouCam2','Reclamacao', 'CustoContato', 'Receita', 'UltimaCampanha']
col=list(df.columns)
"""for i in range (len(col)):
    print(col[i],'=',novo[i])"""
df.rename(columns=dict(zip(col, novo)), inplace=True)

---
#### Verificação de nulos em todas as colunas
---
Coluna 'Renda' tem 24 valores nulos. Os valores nulos serão excluidos. Tabela clientes passou a ter 2216 campos.


In [8]:
df.isna().sum()

AnoNascimento          0
GrauInstrução          0
EstadoCivil            0
Renda                 24
NumCrianca             0
NumAdolescente         0
DataConta              0
UltimaCompra           0
QtdVinho               0
QtdFrutas              0
QtdCarne               0
QtdPeixe               0
QtdDoces               0
QtdOuro                0
QtdComprasDesconto     0
QtdComprasWeb          0
QtdCompraCatalogo      0
QtdCompraLoja          0
NumVisitasSite         0
AceitouCam3            0
AceitouCam4            0
AceitouCam5            0
AceitouCam1            0
AceitouCam2            0
Reclamacao             0
CustoContato           0
Receita                0
UltimaCampanha         0
dtype: int64

In [9]:
df=df.dropna()
df

AnoNascimento GrauInstrução EstadoCivil    Renda  NumCrianca  \
ID                                                                    
5524            1957    Graduation      Single  58138.0           0   
2174            1954    Graduation      Single  46344.0           1   
4141            1965    Graduation    Together  71613.0           0   
6182            1984    Graduation    Together  26646.0           1   
5324            1981           PhD     Married  58293.0           1   
...              ...           ...         ...      ...         ...   
10870           1967    Graduation     Married  61223.0           0   
4001            1946           PhD    Together  64014.0           2   
7270            1981    Graduation    Divorced  56981.0           0   
8235            1956        Master    Together  69245.0           0   
9405            1954           PhD     Married  52869.0           1   

       NumAdolescente   DataConta  UltimaCompra  QtdVinho  QtdFrutas  ...  \
ID                                                                    ...   
5524                0  04-09-2012            58       635         88  ...   
2174                1  08-03-2014            38        11          1  ...   
4141                0  21-08-2013            26       426         49  ...   
6182                0  10-02-2014            26        11          4  ...   
5324                0  19-01-2014            94       173         43  ...   
...               ...         ...           ...       ...        ...  ...   
10870               1  13-06-2013            46       709         43  ...   
4001                1  10-06-2014            56       406          0  ...   
7270                0  25-01-2014            91       908         48  ...   
8235                1  24-01-2014             8       428         30  ...   
9405                1  15-10-2012            40        84          3  ...   

       NumVisitasSite  AceitouCam3  AceitouCam4  AceitouCam5  AceitouCam1  \
ID                                                                          
5524                7            0            0            0            0   
2174                5            0            0            0            0   
4141                4            0            0            0            0   
6182                6            0            0            0            0   
5324                5            0            0            0            0   
...               ...          ...          ...          ...          ...   
10870               5            0            0            0            0   
4001                7            0            0            0            1   
7270                6            0            1            0            0   
8235                3            0            0            0            0   
9405                7            0            0            0            0   

       AceitouCam2  Reclamacao  CustoContato  Receita  UltimaCampanha  
ID                                                                     
5524             0           0             3       11               1  
2174             0           0             3       11               0  
4141             0           0             3       11               0  
6182             0           0             3       11               0  
5324             0           0             3       11               0  
...            ...         ...           ...      ...             ...  
10870            0           0             3       11               0  
4001             0           0             3       11               0  
7270             0           0             3       11               0  
8235             0           0             3       11               0  
9405             0           0             3       11               1  

[2216 rows x 28 columns]

#### Verificação dos valores únicos da coluna 'GraudeInstrução' e tradução

In [10]:
print(df.GrauInstrução.unique())
df['GrauInstrução']= df['GrauInstrução'].str.replace('Graduation', 'Graduado').str.replace('PhD', 'PhD').replace('Master', 'Mestrado').replace('Basic', 'Basico').replace('2n Cycle','Segundo Ciclo' )
print(df.GrauInstrução.unique())

['Graduation' 'PhD' 'Master' 'Basic' '2n Cycle']
['Graduado' 'PhD' 'Mestrado' 'Basico' 'Segundo Ciclo']


#### Verificação dos valoes únicos em 'EstadoCivil' e tradução

In [11]:
print(df.EstadoCivil.unique())
df['EstadoCivil']=df['EstadoCivil'].str.replace('Single', 'Solteiro').replace('Together', 'Juntos').replace('Married','Casado').replace('Divorced', 'Divorciado').replace('Absurd', 'Absurdo').replace('YOLO', 'Só se vive uma vez').replace('Widow', 'Viúvo').replace('Alone', 'Sozinho')
print(df.EstadoCivil.unique())

['Single' 'Together' 'Married' 'Divorced' 'Widow' 'Alone' 'Absurd' 'YOLO']
['Solteiro' 'Juntos' 'Casado' 'Divorciado' 'Viúvo' 'Sozinho' 'Absurdo'
 'Só se vive uma vez']


#### Exclusão das colunas Z_CostContact: CustoContato e Z_Revenue: Receita
Justificativa: essas colunas só tem os valores 3 e 11 repetidos em todas as linhas e não são pertinentes para as analises 

In [12]:
print(df.CustoContato.unique())
print(df.Receita.unique())

[3]
[11]


In [13]:
df=df.drop(columns=['CustoContato', 'Receita'])

In [14]:
df

AnoNascimento GrauInstrução EstadoCivil    Renda  NumCrianca  \
ID                                                                    
5524            1957      Graduado    Solteiro  58138.0           0   
2174            1954      Graduado    Solteiro  46344.0           1   
4141            1965      Graduado      Juntos  71613.0           0   
6182            1984      Graduado      Juntos  26646.0           1   
5324            1981           PhD      Casado  58293.0           1   
...              ...           ...         ...      ...         ...   
10870           1967      Graduado      Casado  61223.0           0   
4001            1946           PhD      Juntos  64014.0           2   
7270            1981      Graduado  Divorciado  56981.0           0   
8235            1956      Mestrado      Juntos  69245.0           0   
9405            1954           PhD      Casado  52869.0           1   

       NumAdolescente   DataConta  UltimaCompra  QtdVinho  QtdFrutas  ...  \
ID                                                                    ...   
5524                0  04-09-2012            58       635         88  ...   
2174                1  08-03-2014            38        11          1  ...   
4141                0  21-08-2013            26       426         49  ...   
6182                0  10-02-2014            26        11          4  ...   
5324                0  19-01-2014            94       173         43  ...   
...               ...         ...           ...       ...        ...  ...   
10870               1  13-06-2013            46       709         43  ...   
4001                1  10-06-2014            56       406          0  ...   
7270                0  25-01-2014            91       908         48  ...   
8235                1  24-01-2014             8       428         30  ...   
9405                1  15-10-2012            40        84          3  ...   

       QtdCompraCatalogo  QtdCompraLoja  NumVisitasSite  AceitouCam3  \
ID                                                                     
5524                  10              4               7            0   
2174                   1              2               5            0   
4141                   2             10               4            0   
6182                   0              4               6            0   
5324                   3              6               5            0   
...                  ...            ...             ...          ...   
10870                  3              4               5            0   
4001                   2              5               7            0   
7270                   3             13               6            0   
8235                   5             10               3            0   
9405                   1              4               7            0   

       AceitouCam4  AceitouCam5  AceitouCam1  AceitouCam2  Reclamacao  \
ID                                                                      
5524             0            0            0            0           0   
2174             0            0            0            0           0   
4141             0            0            0            0           0   
6182             0            0            0            0           0   
5324             0            0            0            0           0   
...            ...          ...          ...          ...         ...   
10870            0            0            0            0           0   
4001             0            0            1            0           0   
7270             1            0            0            0           0   
8235             0            0            0            0           0   
9405             0            0            0            0           0   

       UltimaCampanha  
ID                     
5524                1  
2174                0  
4141                0  
6182                0  
5324                0  
...               ...  
10

In [15]:
df.columns

Index(['AnoNascimento', 'GrauInstrução', 'EstadoCivil', 'Renda', 'NumCrianca',
       'NumAdolescente', 'DataConta', 'UltimaCompra', 'QtdVinho', 'QtdFrutas',
       'QtdCarne', 'QtdPeixe', 'QtdDoces', 'QtdOuro', 'QtdComprasDesconto',
       'QtdComprasWeb', 'QtdCompraCatalogo', 'QtdCompraLoja', 'NumVisitasSite',
       'AceitouCam3', 'AceitouCam4', 'AceitouCam5', 'AceitouCam1',
       'AceitouCam2', 'Reclamacao', 'UltimaCampanha'],
      dtype='object')

In [16]:
df.index.is_unique

True